### PR TITLE
2015 04 08 fh update user

### PR DESCRIPTION
**Pull request by fhartwig**
Don't merge this yet, this branch still needs some cleaning-up and work on tests.

### CODE REVIEW

File: tests/ThentosSpec.hs [hunk 1]
```diff
@@ -69,13 +69,17 @@ spec = do
...
     describe "UpdateUser" $ do
       it "changes user if it exists" $ \ (st, _, _) -> do
-        result <- update' st $ UpdateUser (UserId 1) user1 allowEverything
+        result <- update' st $ UpdateUserField (UserId 1)
+                                               (UpdateUserFieldName "fka_user1")
+                                               allowEverything
         result `shouldBe` Right ()
         result2 <- query' st $ LookupUser (UserId 1) allowEverything
-        result2 `shouldBe` (Right (UserId 1, user1))
+        result2 `shouldBe` (Right (UserId 1, user1 {_userName = "fka_user1"}))
```
Comment (mf59816) [inline tests/ThentosSpec.hs]: or, equally valid:

    result2 `shouldBe` (Right (UserId 1, userName .~ "fka_user1" $ user1))

File: src/Thentos/Smtp.hs [hunk 2]
```diff
@@ -6,6 +6,7 @@ module Thentos.Smtp
     ( sendUserConfirmationMail
     , sendUserExistsMail
     , sendPasswordResetMail
+    , sendEmailChangeConfirmMail
```
Comment (mf59816) [inline src/Thentos/Smtp.hs]: `sendChangeConfirmMail` would be more consistent with the existing names.

File: src/Thentos/Frontend/Handlers.hs [hunk 3]
```diff
@@ -110,20 +111,90 @@ userCreateConfirm = do
...
+                Left NoSuchToken -> blaze $ errorPage "No such token"
+                Left e -> do
+                    logger WARNING (show e)
+                    blaze $ errorPage "Error when trying to change email address"
+
+runWithUserClearance :: (ThentosClearance -> UserId -> FH ()) -> FH ()
 runWithUserClearance = runWithUserClearance' ()
 
-runWithUserClearance' :: a -> (ThentosClearance -> FH a) -> FH a
+runWithUserClearance' :: a -> (ThentosClearance -> UserId -> FH a) -> FH a
```
Comment (mf59816) [inline src/Thentos/Frontend/Handlers.hs]: the fact that the function argument also wants a uid should probably be discussed in a comment.  (answer questions like: how is it filled?  why is this not done before the call to `runWithUserClearance`?)

File: src/Thentos/Frontend/Handlers.hs [hunk 4]
```diff
@@ -110,20 +111,90 @@ userCreateConfirm = do
...
+    (emailView, result) <- runForm "update_email" emailUpdateForm
+    case result of
+        Nothing -> blaze $ userUpdatePage userView emailView passwordView
+        Just newEmail -> do
+            result' <- snapRunAction' clearance $ requestUserEmailChange uid newEmail
+            case result' of
+                Right token -> do
+                    config :: ThentosConfig <- gets (^. cfg)
+                    feConfig <- gets (^. frontendCfg)
+                    liftIO $ sendEmailChangeConfirmMail (Tagged $ config >>. (Proxy :: Proxy '["smtp"]))
```
Comment (mf59816) [inline src/Thentos/Frontend/Handlers.hs]: send out emails in `Thentos.Api`?  the benefit would be that the backend does not have to duplicate that code if email update is every triggered via rest.  (the idea is that thentos can be accessed both via the snap api and -- alternatively -- entirely via rest.)

File: src/Thentos/Frontend.hs [hunk 5]
```diff
@@ -58,7 +58,10 @@ routes = [ ("", ifTop $ H.index)
          , ("user/create_confirm", H.userCreateConfirm)
          , ("user/reset_password_request", H.resetPasswordRequest)
          , ("user/reset_password", H.resetPassword)
-         -- , ("user/update", ?)
+         , ("user/update", H.userUpdate)
+         , ("user/update_email", H.emailUpdate)
+         , ("user/email_confirm", H.emailUpdateConfirm)
```
Comment (mf59816) [inline src/Thentos/Frontend.hs]: `"user/update_email_confirm"` instead?

File: src/Thentos/DB/Trans.hs [hunk 6]
```diff
@@ -244,38 +264,36 @@ trans_resetPassword now token newPass = do
...
 
+-- | Update existing user in DB.  Throw an error if user id does not
+-- exist, or if email address in updated user is already in use by
+-- another user.
 trans_updateUserField :: UserId -> UpdateUserFieldOp -> ThentosUpdate ()
-trans_updateUserField uid op = do
+trans_updateUserField uid op = trans_updateUserFields uid [op]
+
+trans_updateUserFields :: UserId -> [UpdateUserFieldOp] -> ThentosUpdate ()
+trans_updateUserFields uid ops = do
```
Comment (mf59816) [inline src/Thentos/DB/Trans.hs]: nice idea.  this way we can still choose to change any number of fields inside one transaction, but we don't have to.  this makes `updateUser` completely unnecessary.

File: src/Thentos/Smtp.hs (hunk 2 above)
Comment (fhartwig) [inline src/Thentos/Smtp.hs]: How so? It's `send Password-reset mail`, why not `send email-change confirm mail`? (Although it should be "confirmation", not "confirm")

File: tests/ThentosSpec.hs (hunk 1 above)
Comment (fhartwig) [inline tests/ThentosSpec.hs]: Yeah, that's better. I'm really bad at remembering lens operators :disappointed:

File: src/Thentos/Smtp.hs (hunk 2 above)
Comment (mf59816) [inline src/Thentos/Smtp.hs]: you're right.  please don't change it.  (i don't have an opinion on 'confirmation' vs. 'confirm'.)

File: src/Thentos/Frontend.hs (hunk 5 above)
Comment (fhartwig) [inline src/Thentos/Frontend.hs]: :+1:

File: src/Thentos/Frontend/Handlers.hs (hunk 4 above)
Comment (fhartwig) [inline src/Thentos/Frontend/Handlers.hs]: You're right, I'll do that.

File: tests/ThentosSpec.hs (hunk 1 above)
Comment (mf59816) [inline tests/ThentosSpec.hs]: you can always go to the import list in `Thentos.Trans.DB` and then search the module for how the operators mentioned there are used.  that should get you far enough.  (-:

**Comment by mf59816** [comment]
replaced by #28

**Pull request closed without merging**